### PR TITLE
Carry rollout MoE routing from the vLLM sampler to the maxtext trainer

### PR DESCRIPTION
**Pull request by khatwanimohit**
Lets an RL run replay the MoE expert routing a rollout actually used instead of re-running the router during training. Both ends already existed -- vLLM and tpu-inference capture routing (`enable_return_routed_experts` -> `CompletionOutput.routed_experts`, `[length, num_layers, top_k]`) and MaxText consumes it (`forced_routed_experts`) -- so this is the middle.

Sampler side:
- `VllmConfig.return_routed_experts` sets vLLM's engine arg; `VllmSampler.detokenize` collects the arrays and `SamplerOutput` carries them.
- The in-process adapter populates `SamplingResponse.routed_experts`, and warns when a request asks for routing the engine was not built to capture, since capture is engine-level and the alternative is silently handing the trainer None -- indistinguishable from a dense model.

Pipeline side:
- `routed_experts` on `TokenSegment`, `TrajectoryItem` and `RLTrainerPayload`, plus `UNSET_ROUTED_EXPERT = -1` for slots the trainer must not replay.
- The GRPO adapter pads short captures with the sentinel rather than expert 0.
- The padded assembler lays routing out over `[prompt | completion]` the same way the token ids are padded: prompts right-aligned keeping the tail (as `_left_pad` does), completions left-aligned. Getting this wrong detaches every replayed expert from its token -- silent, not a crash.

Demo side (math_gsm8k_dist), all opt-in:
- trainer node: `--trainer_engine=maxtext` swaps PeftTrainer for MaxText's engine wired for replay, importing MaxText lazily and skipping the tunix actor model, which only knows the demo's dense architectures. `--maxtext_load_parameters_path` points it at the checkpoint; without one it warns, since random trainer weights make replayed routing meaningless.
- rollout node: `--return_routed_experts`, and `--load_rollout_weights` to load the real checkpoint instead of vLLM's dummy weights. With both sides on the same weights a first step is meaningful with no weight transfer at all.
- `launcher.sh` and `k8s_launcher.sh`: `ROUTER_REPLAY=1` sets the above; slices and meshes are now configurable rather than hardcoded.
- Dockerfile: opt-in `INSTALL_MAXTEXT` build arg, which also drops triton (GPU-only, and its native module aborts the process when MaxText imports on a TPU host).

Tests pin layout rather than mere presence, since the realistic failure is misalignment. The loss assertions are differential -- two different replays must disagree -- because a "loss is finite" check passes just as happily when the routing is dropped. Each was verified to fail against a mutation that disables what it covers.

**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

### CODE REVIEW

File: tests/experimental/worker/router_replay_trainer_worker_test.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+      learning_rate=1e-2,
+      warmup_steps_fraction=0.0,
+  )
+
+
+def _payload(forced):
+  """Wraps a forced-routing array in a payload the replay loss fn accepts."""
+  forced[:, -PAD_LEN:] = datatypes.UNSET_ROUTED_EXPERT
+  token_ids, token_mask = _tokens()
+  return maxtext_engine.RouterReplayTrainerPayload(
```
Comment (tianshub) [inline tests/experimental/worker/router_replay_trainer_worker_test.py]: Since the field is added to [RLTrainerPayload](https://github.com/google/tunix/blob/795635a28430c03de2911a5c595db59748e3e2bf/tunix/experimental/common/datatypes.py#L535), why do you still need to use RouterReplayTrainerPayload here? The actual data class that passed to maxtext should be RLTrainerPayload no?

File: tunix/experimental/common/datatypes.py [hunk 2]
```diff
@@ -289,6 +302,16 @@ def __post_init__(self):
       raise ValueError(
           f"logps shape {self.logps.shape} != tokens shape {self.tokens.shape}"
       )
+    # Only the leading (per-token) axis has to line up; the trailing axes are
+    # [num_layers, top_k] and are model-dependent.
+    if (
+        self.routed_experts is not None
+        and self.routed_experts.shape[0] != self.tokens.shape[0]
```
Comment (tianshub) [inline tunix/experimental/common/datatypes.py]: you can add a check on the ndim=3 as well

File: tunix/experimental/examples/math_gsm8k_dist/k8s_launcher.sh [hunk 3]
```diff
@@ -46,6 +46,17 @@ export WEIGHT_SYNC_MODE=${WEIGHT_SYNC_MODE:-none}
 # MaxText trainer configuration: only consulted when TRAINER_BACKEND=maxtext
 export MAXTEXT_MODEL_NAME=${MAXTEXT_MODEL_NAME:-qwen3-1.7b}
 export MAXTEXT_CKPT=${MAXTEXT_CKPT:-}
+# Router replay: the rollout records the MoE experts it routed through and the
+# MaxText trainer replays them instead of re-running its own gate. MoE models
+# only; requires TRAINER_BACKEND=maxtext.
+#
+# This also loads the real checkpoint into the rollout instead of vLLM's dummy
```
Comment (tianshub) [inline tunix/experimental/examples/math_gsm8k_dist/k8s_launcher.sh]: put the comment after L58?

File: tunix/experimental/examples/math_gsm8k_dist/launcher.sh [hunk 4]
```diff
@@ -55,6 +55,17 @@ WANDB_RUN_NAME=${WANDB_RUN_NAME:-}
 WANDB_API_KEY=${WANDB_API_KEY:-}
 SAMPLER=${SAMPLER:-inprocess_vllm}
 WEIGHT_SYNC_MODE=${WEIGHT_SYNC_MODE:-none}
+# Router replay: the rollout records the MoE experts it routed through and the
```
Comment (tianshub) [inline tunix/experimental/examples/math_gsm8k_dist/launcher.sh]: we can delete this comment since launch script already explained it

File: tunix/experimental/examples/math_gsm8k_dist/launcher.sh [hunk 5]
```diff
@@ -55,6 +55,17 @@ WANDB_RUN_NAME=${WANDB_RUN_NAME:-}
...
+# MaxText trainer replays them instead of re-running its own gate. MoE models
+# only; both must be on for replay to happen.
+ROUTER_REPLAY=${ROUTER_REPLAY:-0}
+TRAINER_BACKEND=${TRAINER_BACKEND:-tunix}
+MAXTEXT_MODEL_NAME=${MAXTEXT_MODEL_NAME:-qwen3.5-35b-a3b}
+MAXTEXT_CKPT=${MAXTEXT_CKPT:-}
+# Load the real checkpoint into vLLM instead of its dummy random weights, so
```
Comment (tianshub) [inline tunix/experimental/examples/math_gsm8k_dist/launcher.sh]: ditto

File: tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py [hunk 6]
```diff
@@ -275,6 +285,69 @@ def close(self) -> None:
...
+  `forced_routed_experts`.
+  """
+  from maxtext.training_engine import maxtext_engine  # pylint: disable=g-import-not-at-top
+
+  def gen_model_input_fn(payload: Any) -> dict[str, Any]:
+    token_ids, token_mask = _concat_prompt_completion(payload)
+    # Reuse MaxText's adapter for positions/targets/segmentation rather than
+    # re-deriving them; it already handles left-padded rows and the roll(-1)
+    # wrap-around at the sequence end.
+    shim = maxtext_engine.RouterReplayTrainerPayload(
```
Comment (tianshub) [inline tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py]: this redirection shouldn't be necessary right? the router_replay_gen_model_input_fn can directly consume RLTrainerPayload and extract the fields

File: tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py [hunk 7]
```diff
@@ -275,6 +285,69 @@ def close(self) -> None:
       self._trainer.close()
 
 
+def _mask_or_ones(mask: Any, ids: np.ndarray) -> np.ndarray:
+  """`mask` as an array, or an all-ones mask matching `ids` when absent."""
+  return np.ones_like(ids) if mask is None else np.asarray(mask)
+
+
+def _concat_prompt_completion(payload: Any) -> tuple[np.ndarray, np.ndarray]:
```
Comment (tianshub) [inline tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py]: we should reduce these kind of functions from the script as much as possible and just bake it into the framework, let's put it in the router_replay_gen_model_input_fn. Also I doubt if you need to have separate gen_model_input function for this, wouldn't it be easier to have one function that can deal with optional field?

File: tunix/experimental/orchestrator/batch_assembly.py [hunk 8]
```diff
@@ -54,6 +54,36 @@ def _left_pad(
   return out, mask
 
 
+def _routed_experts_aligned(
+    routed: np.ndarray,
```
Comment (tianshub) [inline tunix/experimental/orchestrator/batch_assembly.py:58]: add comment for shape annotation

File: tunix/experimental/orchestrator/batch_assembly.py [hunk 9]
```diff
@@ -54,6 +54,36 @@ def _left_pad(
...
+) -> np.ndarray:
+  """Lays `[P + C, num_layers, top_k]` routing out over the padded `[P | C]`.
+
+  Mirrors how the token ids themselves are padded -- prompts right-aligned in
+  the prompt window (keeping the tail, as `_left_pad` does) and completions
+  left-aligned in the response window -- so replayed routing stays attached to
+  the token it was captured for. Everything else is left unset.
+  """
+  routed = np.asarray(routed, dtype=np.int32)
+  prompt_part = routed[:prompt_len][-max_prompt_length:]
```
Comment (tianshub) [inline tunix/experimental/orchestrator/batch_assembly.py]: I'm a bit confused on the routed shape, isn't dim0 suppose to be P + C? why are we slicing on dim0 and dim1?

**Review by tianshub — CHANGES_REQUESTED** [review]

File: tests/experimental/worker/router_replay_trainer_worker_test.py (hunk 1 above)
Comment (khatwanimohit) [reply tests/experimental/worker/router_replay_trainer_worker_test.py]: `RLTrainerPayload` is the only payload involved now

File: tunix/experimental/common/datatypes.py (hunk 2 above)
Comment (khatwanimohit) [reply tunix/experimental/common/datatypes.py]: Done. `TokenSegment.__post_init__` now rejects a non-3D `routed_experts` with the expected `[length, num_layers, top_k]` in the message

File: tunix/experimental/examples/math_gsm8k_dist/k8s_launcher.sh (hunk 3 above)
Comment (khatwanimohit) [reply tunix/experimental/examples/math_gsm8k_dist/k8s_launcher.sh]: done

File: tunix/experimental/examples/math_gsm8k_dist/launcher.sh (hunk 4 above)
Comment (khatwanimohit) [reply tunix/experimental/examples/math_gsm8k_dist/launcher.sh]: Deleted

File: tunix/experimental/examples/math_gsm8k_dist/launcher.sh (hunk 5 above)
Comment (khatwanimohit) [reply tunix/experimental/examples/math_gsm8k_dist/launcher.sh]: Deleted

File: tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py (hunk 6 above)
Comment (khatwanimohit) [reply tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py]: done

File: tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py (hunk 7 above)
Comment (khatwanimohit) [reply tunix/experimental/examples/math_gsm8k_dist/run_trainer_node.py]: agreed

File: tunix/experimental/orchestrator/batch_assembly.py (hunk 9 above)
Comment (khatwanimohit) [reply tunix/experimental/orchestrator/batch_assembly.py]: made it clearer now

File: tunix/experimental/orchestrator/batch_assembly.py (hunk 8 above)
Comment (khatwanimohit) [reply tunix/experimental/orchestrator/batch_assembly.py:58]: done

File: tunix/rl/common.py [hunk 10]
```diff
@@ -337,6 +410,7 @@ def compute_per_token_logps(
     segment_positions: jax.Array | None = None,
     temperature: float = 1.0,
     chunk_size: int = 0,
+    forced_routed_experts: jax.Array | None = None,
```
Comment (tianshub) [inline tunix/rl/common.py]: should this just be routed_experts so it's consistent with others?

**Review by tianshub — APPROVED** [review]
Thank you for the update, looks much cleaner.

File: tunix/rl/common.py (hunk 10 above)
Comment (khatwanimohit) [reply tunix/rl/common.py]: changed to routed_experts
